resolve.go

```diff
@@ -104,179 +106,162 @@ func (r *Resolver) getResolver(domain string) BasicResolver {
 	return r.def
 }
 
-// Resolve resolves a DNS multiaddr.
+// Resolve resolves a DNS multiaddr. It will only resolve the first DNS component in the multiaddr.
+// If you need to resolve multiple DNS components, you may call this function again with each returned address.
 func (r *Resolver) Resolve(ctx context.Context, maddr ma.Multiaddr) ([]ma.Multiaddr, error) {
-	var results []ma.Multiaddr
-	for i := 0; maddr != nil; i++ {
-		var keep ma.Multiaddr
-
-		// Find the next dns component.
-		keep, maddr = ma.SplitFunc(maddr, func(c ma.Component) bool {
-			switch c.Protocol().Code {
-			case dnsProtocol.Code, dns4Protocol.Code, dns6Protocol.Code, dnsaddrProtocol.Code:
-				return true
-			default:
-				return false
-			}
-		})
-
-		// Keep everything before the dns component.
-		if keep != nil {
-			if len(results) == 0 {
-				results = []ma.Multiaddr{keep}
-			} else {
-				for i, r := range results {
-					results[i] = r.Encapsulate(keep)
-				}
-			}
+	// Find the next dns component.
+	preDNS, maddr := ma.SplitFunc(maddr, func(c ma.Component) bool {
```

Comment:
Should we ensure that `preDNS` is empty? 

All dns multiaddrs MUST begin with `/dnsXXX` right?

Reply:
In other words, we should only resolve a dns address at the front of the multiaddr. If someone wants to resolve something after the first component, they can split the multiaddr, use this method, stitch the multiaddr back.

Reply:
(Discussed out of band)

This library commonly assumes that the DNS component can be anywhere. This change is in line with that assumption. A reasonable way to use `Resolve` is to continually resolve a multiaddr until `Matches` returns false. Since `Matches` returns true if any part of the multiaddr contains a DNS component, it makes sense that Resolve should resolve a DNS component anywhere in the multiaddr.

Will keep this change as is.